### PR TITLE
Implementation of Throw Parsing

### PR DESCRIPTION
**Pull request by matjin**

### CODE REVIEW

File: Cilsil/Cil/Parsers/ThrowParser.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                                                    state.ProcDesc);
+                    Expression returnVariable = new LvarExpression(
+                        new LocalVariable(Identifier.ReturnIdentifier,
+                                          state.Method));
+                    (var returnValue, _) = state.Pop();
+                    var retInstr = new Store(returnVariable,
+                                             new ExnExpression(returnValue),
+                                             Typ.FromTypeReference(retType),
+                                             state.CurrentLocation);
+                    AddMethodBodyInstructionsToCfg(state, retInstr);
```
Comment (xi-liu-ds) [inline Cilsil/Cil/Parsers/ThrowParser.cs]: It seems that `throw` instruction not only creating a `ret` expression but also a `store` expression.
image

File: Cilsil/Cil/Parsers/ThrowParser.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    internal class ThrowParser : InstructionParser
+    {
+        protected override bool ParseCilInstructionInternal(Instruction instruction,
+                                                            ProgramState state)
+        {
+            switch (instruction.OpCode.Code)
+            {
+                case Code.Throw:
+                    var retType = state.Method.ReturnType.GetElementType();
+                    var retNode = new StatementNode(state.CurrentLocation,
```
Comment (xi-liu-ds) [inline Cilsil/Cil/Parsers/ThrowParser.cs:22]: Why this node never used or added to CFG?

**Review by xi-liu-ds — CHANGES_REQUESTED** [review]

File: Cilsil/Cil/Parsers/ThrowParser.cs (hunk 1 above)
Comment (matjin) [reply Cilsil/Cil/Parsers/ThrowParser.cs]: Store is an instruction, not an expression -- which is what we add to the node 

                    var retInstr = new Store(returnVariable,
                                             new ExnExpression(returnValue),
                                             Typ.FromTypeReference(retType),
                                             state.CurrentLocation);
                    retNode.Instructions.Add(retInstr);

File: Cilsil/Cil/Parsers/ThrowParser.cs (hunk 2 above)
Comment (matjin) [reply Cilsil/Cil/Parsers/ThrowParser.cs:22]: Forgot -- removed it thanks!

File: Cilsil/Cil/Parsers/ThrowParser.cs (hunk 1 above)
Comment (xi-liu-ds) [reply Cilsil/Cil/Parsers/ThrowParser.cs]: I mean the `n$3=*&irvar0:java.io.IOException*` instruction in the screenshot. It seems that you didn't create this instruction and add to the return node.

Comment (matjin) [reply Cilsil/Cil/Parsers/ThrowParser.cs]: This is accomplished by a previous instruction, as the [official documentation](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.throw?view=net-5.0) describes that the object reference is already pushed onto the stack prior:

"An object reference (to an exception) is pushed onto the stack.

The object reference is popped from the stack and the exception thrown."

For example:
    IL_0001:  newobj     instance void [System.Runtime]System.IO.FileNotFoundException::.ctor()
    IL_0006:  throw

**Review by xi-liu-ds — APPROVED** [review]

**Pull request merged**